### PR TITLE
Add optional server shutdown delay + announcement

### PR DESCRIPTION
**Pull request by morrissimo**
My first Go! 😅

* adds: new argument `StopServerAnnounceDelay` of type time.Duration - defaults to `0s`
* adds: new helper method `announceStopViaConsole()` - my attempt at string formatting 👍 
* adds: logic to announce server shutdown if argument `StopServerAnnounceDelay` > 0, then `time.Sleep(StopServerAnnounceDelay)` 
* adds: few more log statements

Please tweak as you see fit, or let me know what you'd like to see changed to better reflect whatever vision you already had for this capability. Thanks for such a great project!

Refs https://github.com/itzg/mc-server-runner/issues/10

### CODE REVIEW

**Comment by morrissimo** [comment]
test results:

```
$ docker run -it --rm \
  -v ${PWD}:/build \
  -w /build \
  circleci/golang:1.12
$ go run main.go test/dump.sh
HOSTNAME=608e02ea8cf1
HOME=/home/circleci
TERM=xterm
PATH=/home/circleci/.local/bin:/home/circleci/bin:/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
LANG=C.UTF-8
DEBIAN_FRONTEND=noninteractive
GOPATH=/go
PWD=/build
GOLANG_VERSION=1.12.17
2021-07-29T01:03:18.706Z        INFO    mc-server-runner        Done
$ go run main.go test/bash-only.sh
test/bash-only.sh: 3: test/bash-only.sh: [[: not found
2021-07-29T01:04:56.734Z        INFO    mc-server-runner        Done
$ go run main.go --shell sh test/bash-only.sh
test/bash-only.sh: 3: test/bash-only.sh: [[: not found
2021-07-29T01:05:15.256Z        INFO    mc-server-runner        Done
```

**Comment by itzg** [comment]
The changes look great! 

As for the failing test case, I'm thinking I either documented that wrong and `--shell bash` is needed or the shebang line needs to be `#!/bin/bash` here

https://github.com/itzg/mc-server-runner/blob/823ee3c58763b6545c2077ec5572d6d383fb7dd1/test/bash-only.sh#L1

Speaking of which, several of these features are tough to test. Have you been able to test yours out with some manual approach?

File: main.go [hunk 1]
```diff
@@ -200,6 +207,14 @@ func stopWithRconCli() error {
 	return rconCliCmd.Run()
 }
 
+func announceStopViaConsole(logger *zap.Logger, stdin io.Writer, shutdownDelay time.Duration) {
+	logger.Info("Sending shutdown announce 'say' to Minecraft server...")
+	_, err := stdin.Write([]byte(fmt.Sprintf("say Server shutting down in %0.f seconds\n", shutdownDelay.Seconds())))
```
Comment (itzg) [inline main.go:212]: Floating format might look overly "precise". How about just integer formatting with

```suggestion
	_, err := stdin.Write([]byte(fmt.Sprintf("say Server shutting down in %d seconds\n", shutdownDelay.Seconds())))
```

Comment (morrissimo) [reply main.go:212]: I tried int formatting initially - but the `.Seconds()` method returns a float. The `%0.f` format code specifies 0 decimal precision for floats, so the effect ends up the same as `%d`

* Go playground https://play.golang.org/p/zw-EkgTPoSL

**Comment by morrissimo** [comment]
> Speaking of which, several of these features are tough to test. Have you been able to test yours out with some manual approach?

No, I haven't tried anything more than verifying that the new argument appeared to be implemented correctly by running `go run main.go --help`. Any suggestions on approaches?

File: main.go (hunk 1 above)
Comment (itzg) [reply main.go:212]: Aha, I mentioned that without thinking what the return type was.

**Comment by itzg** [comment]
Oh I remember now: using an existing Minecraft server data directory I would test by running `java -jar server.jar` via mc server runner. Then Control-C (if you're not on Windows) should trigger the signal handling.

**Comment by morrissimo** [comment]
> Oh I remember now: using an existing Minecraft server data directory I would test by running `java -jar server.jar` via mc server runner. Then Control-C (if you're not on Windows) should trigger the signal handling.

(Finally got some time to get back to this)

After figuring out how to build a compatible binary, I downloaded a compatible vanilla Minecraft server jar [1] and ran it to get through the initial server setup, which ran successfully - `eula.txt` lol, memories

Then I ran the modified version of mc-server-runner with the following command (not using the new arg):
```
./mc-server-runner java -jar server.jar --nogui
```
...but CTRL-C didn't lead to any of the log output that it seemed like it should (having seen the `main.go` source code): mc-server-runner simply exited immediately

Unsurprisingly, trying with the new command line arg produced the same behavior:
```
./mc-server-runner -stop-server-announce-delay 10s java -jar server.jar --nogui
```
...but CTRL-C again led to an immediate exit

Am I using the correct command line?

**Comment by morrissimo** [comment]
Tried with the current `master` version as well, just to be sure - same behavior: CTRL-C leads to an immediate exit with no info log output

**Comment by morrissimo** [comment]
ah - CTRL-C sends SIGINT but the handler is only listening for SIGTERM! Tested it again with a `kill <pid>` (which sends SIGTERM) and saw all the logging output I was expecting

Will test a bit more then probably a few more tweaks to push here

**Comment by morrissimo** [comment]
without new argument:
```
./mc-server-runner java -jar server.jar
2021-07-30T22:26:52.001-0400	INFO	mc-server-runner	startup
# [usual startup output snipped]
[22:27:12] [Server thread/INFO]: Done (4.367s)! For help, type "help"
[22:27:12] [Server thread/INFO]: Starting GS4 status listener
[22:27:17] [Server thread/INFO]: Thread Query Listener started
[22:27:17] [Query Listener #1/INFO]: Query running on 0.0.0.0:25565
2021-07-30T22:27:37.921-0400	INFO	mc-server-runner	Sending 'stop' to Minecraft server...
2021-07-30T22:27:37.921-0400	INFO	mc-server-runner	Waiting for completion...
[22:27:37] [Server thread/INFO]: Stopping the server
[22:27:38] [Server thread/INFO]: Stopping server
[22:27:38] [Server thread/INFO]: Saving players
[22:27:38] [Server thread/INFO]: Saving worlds
# [usual shutdown output snipped]
[22:27:38] [Server thread/INFO]: Thread Query Listener stopped
2021-07-30T22:27:40.632-0400	INFO	mc-server-runner	Done
```

with new argument:
```
$ ./mc-server-runner -stop-server-announce-delay 10s java -jar server.jar
2021-07-30T22:28:20.999-0400	INFO	mc-server-runner	startup
# [usual startup output snipped]
[22:28:43] [Server thread/INFO]: Done (4.873s)! For help, type "help"
[22:28:43] [Server thread/INFO]: Starting GS4 status listener
[22:28:48] [Server thread/INFO]: Thread Query Listener started
[22:28:48] [Query Listener #1/INFO]: Query running on 0.0.0.0:25565
2021-07-30T22:28:56.074-0400	INFO	mc-server-runner	Sending shutdown announce 'say' to Minecraft server
2021-07-30T22:28:56.074-0400	INFO	mc-server-runner	Sleeping before server stop	{"sleepTime": "10s"}
[22:28:56] [Server thread/INFO]: [Server] Server shutting down in 10 seconds
2021-07-30T22:29:06.077-0400	INFO	mc-server-runner	Sending 'stop' to Minecraft server...
2021-07-30T22:29:06.077-0400	INFO	mc-server-runner	Waiting for completion...
[22:29:06] [Server thread/INFO]: Stopping the server
[22:29:06] [Server thread/INFO]: Stopping server
[22:29:06] [Server thread/INFO]: Saving players
[22:29:06] [Server thread/INFO]: Saving worlds
# [usual shutdown output snipped]
[22:29:07] [Server thread/INFO]: Thread Query Listener stopped
2021-07-30T22:29:08.669-0400	INFO	mc-server-runner	Done
```

**Review by itzg — APPROVED** [review]
Thanks! I'm going to get this merged and tagged. Then I'll add an option to the server image to use it.

**Comment by ChipWolf** [comment]
@morrissimo an enhancement request here: make the command(s) totally configurable.
Perhaps read a 'shutdown' file at the root so we can use a set of commands

Thanks

**Pull request merged**